### PR TITLE
API-2532: Added atomicSave to SubscriptionFieldsRepository to ensure …

### PR DESCRIPTION
**Pull request by googley42**
…save and retrieval of data is atomic.

### CODE REVIEW

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 1]
```diff
@@ -19,20 +19,25 @@ package uk.gov.hmrc.apisubscriptionfields.repository
...
 import uk.gov.hmrc.mongo.ReactiveRepository
 import uk.gov.hmrc.mongo.json.ReactiveMongoFormats
 
+import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
+
 @ImplementedBy(classOf[SubscriptionFieldsMongoRepository])
 trait SubscriptionFieldsRepository {
+  type IsInsert = Boolean
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Why did you create this alias? 
What is the purpose of it?

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 2]
```diff
@@ -75,8 +80,34 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
     )
   )
 
-  override def save(subscription: SubscriptionFields): Future[(SubscriptionFields, Boolean)] = {
-    save(subscription, selectorForSubscriptionFields(subscription))
+  override def saveAtomic(subscription: SubscriptionFields): Future[(SubscriptionFields, IsInsert)] = {
+
+    val selector = selectorForSubscriptionFields(subscription)
+    val entity = subscription
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: it is not needed to create this `entity` variable
you can directly use `subscription`

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 1 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: It it to make the code more readable by revealing the intention of the type. eg below we can more easily see why `!updateLastError.updatedExisting` is being inverted
``` scala
    mongoCollection.findAndModify(selector, updateOp).map {findAndModifyResult =>
      val maybeTuple: Option[(SubscriptionFields, IsInsert)] = for {
        value <- findAndModifyResult.value
        updateLastError <- findAndModifyResult.lastError
      } yield (value.as[SubscriptionFields], !updateLastError.updatedExisting)
```

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 2 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: fixed

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 1 above)
Comment (gokyo) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Ok.
It makes sense. The code is more readable.
I would ask you two things:
1) Could you please move this alias to the model?
2) Could you replace `Boolean` with this alias in the relevant code for all layers (repository, service, controller) for both field definitions and subscription fields?

Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: I have pushed extended scope of this improvement

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 3]
```diff
@@ -75,8 +78,33 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
...
+    mongoCollection.findAndModify(selector, updateOp).map { findAndModifyResult =>
+      val maybeTuple: Option[(SubscriptionFields, IsInsert)] = for {
+        value <- findAndModifyResult.value
+        updateLastError <- findAndModifyResult.lastError
+      } yield (value.as[SubscriptionFields], !updateLastError.updatedExisting)
+
+      maybeTuple.fold[(SubscriptionFields, IsInsert)] {
+        val error = s"Error upserting database for $selector."
+        Logger.error(s"$error lastError: ${findAndModifyResult.lastError}")
+        throw new RuntimeException(error)
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Could you please add a unit test covering this exception being thrown?

**Pull request merged**